### PR TITLE
expression, executor: rewrite builtin strcmp function

### PR DESCRIPTION
**Pull request by SteveZhangBit**

### CODE REVIEW

File: expression/builtin_string_test.go [hunk 1]
```diff
@@ -382,33 +382,42 @@ func (s *testEvaluatorSuite) TestReverse(c *C) {
...
-		{[]interface{}{nil, ""}, nil},
+		{[]interface{}{"123", "123"}, false, false, 0},
+		{[]interface{}{"123", "1"}, false, false, 1},
+		{[]interface{}{"1", "123"}, false, false, -1},
+		{[]interface{}{"123", "45"}, false, false, -1},
+		{[]interface{}{123, "123"}, false, false, 0},
+		{[]interface{}{"12.34", 12.34}, false, false, 0},
+		{[]interface{}{nil, "123"}, true, false, 0},
+		{[]interface{}{"123", nil}, true, false, 0},
+		{[]interface{}{"123", errors.New("must err")}, false, true, 0},
```
Comment (SteveZhangBit) [inline expression/builtin_string_test.go]: Added new test cases .

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

File: expression/builtin_string.go [hunk 2]
```diff
@@ -574,35 +574,38 @@ type strcmpFunctionClass struct {
...
-	}
-	left, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinStrcmpSig) evalInt(row []types.Datum) (int64, bool, error) {
+	var left, right string
+	var isNull bool
+	var err error
+
+	left, isNull, err = b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: extract 
sc := b.ctx.GetSessionVars().StmtCtx

thus we can use it in line 603

File: expression/builtin_string.go [hunk 3]
```diff
@@ -574,35 +574,38 @@ type strcmpFunctionClass struct {
...
-		return types.Datum{}, errors.Trace(err)
-	}
-	if args[0].IsNull() || args[1].IsNull() {
-		return d, nil
-	}
-	left, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinStrcmpSig) evalInt(row []types.Datum) (int64, bool, error) {
+	var left, right string
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: var(
    left, right string
    isNull bool
    err error
)

**Comment by SteveZhangBit** [comment]
@XuHuaiyu 
Changed code style.

File: expression/builtin_string.go [hunk 4]
```diff
@@ -574,35 +574,41 @@ type strcmpFunctionClass struct {
 }
 
 func (c *strcmpFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinStrcmpSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeLonglong)
+	tp.Flen = 2
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: What does it mean? @XuHuaiyu

Comment (XuHuaiyu) [inline expression/builtin_string.go]: the `byte length` of the function result 
when be represented as `string`.

File: expression/builtin_string.go [hunk 5]
```diff
@@ -574,35 +574,41 @@ type strcmpFunctionClass struct {
...
+	sig := &builtinStrcmpSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinStrcmpSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
 // eval evals a builtinStrcmpSig.
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: When you change a function name, don't forget to update the comment, here it should begin with `evalInt ...`
@SteveZhangBit

Comment (SteveZhangBit) [inline expression/builtin_string.go]: Understood, I will push it later.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**